risc0/zkvm/src/host/server/exec/executor.rs

```diff
@@ -246,28 +276,46 @@ impl<'a> ExecutorImpl<'a> {
 
     /// Run the executor until [ExitCode::Paused] or [ExitCode::Halted] is
     /// reached, producing a [Session] as a result.
+    #[tracing::instrument(level = "trace", skip_all, err)]
     pub fn run_with_callback<F>(&mut self, callback: F) -> Result<Session, ExecutorError>
     where
         F: FnMut(Segment) -> Result<Box<dyn SegmentRef>>,
     {
-        let mut guest_session = match self.run_guest_only_with_callback(callback) {
-            Ok(session) => session,
-            Err(e) => return Err(ExecutorError::Error(e)),
-        };
-        match guest_session.exit_code {
+        let start = std::time::Instant::now();
+        let mut guest_session = self
+            .run_guest_only_with_callback(callback)
+            .map_err(ExecutorError::from)?;
+        let result = match guest_session.exit_code {
             ExitCode::Fault => {
-                let fault_checker_session = match self.run_fault_checker() {
-                    Ok(session) => session,
-                    Err(e) => return Err(ExecutorError::Error(e)),
-                };
-                for segment in fault_checker_session.segments {
-                    guest_session.segments.push(segment);
-                }
-                guest_session.journal = fault_checker_session.journal;
-                Err(ExecutorError::Fault(guest_session))
+                let Session {
+                    segments, journal, ..
+                } = self.run_fault_checker().map_err(ExecutorError::from)?;
+                segments
+                    .into_iter()
+                    .for_each(|s| guest_session.segments.push(s));
+                guest_session.journal = journal;
+                Err(guest_session.into())
             }
             _ => Ok(guest_session),
+        };
+
+        if let Ok(ref session) = result {
+            let execution_stats = ExecutionStats {
+                total_cycles: self.total_cycles(),
+                session_cycles: self.session_cycle(),
+                segment_count: session.segments.len(),
+                execution_time: start.elapsed(),
+            };
+
+            tracing::debug!(total_cycles = ?execution_stats.total_cycles);
+            tracing::debug!(session_cycles = ?execution_stats.session_cycles);
+            tracing::debug!(segment_count = ?execution_stats.segment_count);
+            tracing::debug!(execution_time = ?execution_stats.execution_time);
+            tracing::debug!(raw_execution_stats = ?execution_stats);
+            tracing::debug!(%execution_stats);
```

Comment:
Added redundant traces in different lines so users can filter exactly what they want to show in the logs. 

In order to print this information selectively, it's necessary to set `RUST_LOG` in accordance with [Tracing Subscriber Filter Directives](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/filter/struct.EnvFilter.html#directives)

Example: 

```bash
RUST_LOG=[{execution_time}],[{total_cycles}],[{session_cycles}],[{segment_count}],[{raw_execution_stats}],[{execution_stats}]
```

Although we have the flexibility to print those logs only selecting specific fields, the recommended way for users to get the information would be to use only the `execution_stats` directive: 

```bash
$ RUST_LOG=[{execution_stats}] cargo run --release

2023-10-27T18:43:24.604781Z DEBUG run_with_callback: risc0_zkvm::host::server::exec::executor: execution_stats=Execution Statistics:
Total Cycles: 33981
Session Cycles: 23252
Segment Count: 1
Execution Time: 3.492208ms
```

Reply:
@mothran , @SchmErik , I'd like to ask your opinion on the data retrieval, as I'm not sure if I'm obtaining it appropriately. 

Could you please confirm or suggest changes?

Reply:
I'd like to avoid using `tracing::debug` until we consistently do so across the entire codebase in one NFC PR. For now, would it be OK to just use `log::debug`?

Reply:
Oh definitely! Thanks for the quick reply, I'll change it now ✅

Reply:
@flaub , I changed from `tracing` to `log` (and `tracing_subscriber` to `env_logger`). 

It seems we lose the ability to filter the logs more accurately, though. 
What I did was to add a prefix for the logs we want to show, I'm not aware of a better way to do it.